src/zpiv/zpivmodule.cpp

```diff
@@ -39,29 +86,40 @@ const uint256 PublicCoinSpend::signatureHash() const
 
 namespace ZPIVModule {
 
-    bool createInput(CTxIn &in, CZerocoinMint &mint, uint256 hashTxOut) {
-        libzerocoin::ZerocoinParams *params = Params().Zerocoin_Params(false);
-        uint8_t nVersion = mint.GetVersion();
-        if (nVersion < libzerocoin::PrivateCoin::PUBKEY_VERSION) {
-            // No v1 serials accepted anymore.
-            return error("%s: failed to set zPIV privkey mint version=%d\n", __func__, nVersion);
+    bool createInput(CTxIn &in, CZerocoinMint &mint, uint256 hashTxOut, const int spendVersion) {
+        bool fUseV1Params = mint.GetVersion() < libzerocoin::PrivateCoin::PUBKEY_VERSION;
+        libzerocoin::ZerocoinParams *params = Params().Zerocoin_Params(fUseV1Params);
+        if (fUseV1Params && spendVersion < PUBSPEND_SCHNORR) {
+            // v1 coins need at least version 4 spends
+            std::string errMsg = strprintf("Unable to create PublicCoinSpend version %d with coin version 1. "
+                    "Minimum spend version required: %d", spendVersion, PUBSPEND_SCHNORR);
+            return error("%s: %s", __func__, errMsg);
         }
 
         CKey key;
-        if (!mint.GetKeyPair(key))
-            return error("%s: failed to set zPIV privkey mint version=%d\n", __func__, nVersion);
+        CPubKey pk;
+        CPubKey* pk_ptr = nullptr;
+        if (!fUseV1Params) {
+            if (!mint.GetKeyPair(key))
+                return error("%s: failed to set zPIV privkey mint.", __func__);
+            pk = key.GetPubKey();
+            pk_ptr = &pk;
+        }
+
+        PublicCoinSpend spend(params, spendVersion, mint.GetSerialNumber(),
+                mint.GetRandomness(), hashTxOut, pk_ptr);
```

Comment:
You can move this constructor above the key creation, left pk_ptr as nullptr and then join both `if(!fUseV1Params)` below, generate the pk and set it, same as we are setting the rest of the params.

Reply:
Absoutely :thumbsup:
in https://github.com/PIVX-Project/PIVX/pull/936/commits/d71e2235482b3a63af69679cda31470aebe1d00f (added needed setter in `CoinSpend.h`)